knn-banking-product-recommender/training/data_prep.py

```diff
@@ -0,0 +1,170 @@
+        "ind_dela_fin_ult1": "long_term_deposits",
+        "ind_ecue_fin_ult1": "e_acc",
+        "ind_fond_fin_ult1": "funds",
+        "ind_hip_fin_ult1": "mortgage",
+        "ind_plan_fin_ult1": "pensions_plan",
+        "ind_pres_fin_ult1": "loans",
+        "ind_reca_fin_ult1": "taxes",
+        "ind_tjcr_fin_ult1": "credit_card",
+        "ind_valo_fin_ult1": "securities",
+        "ind_viv_fin_ult1": "home_acc",
+        "ind_nomina_ult1": "payroll",
+        "ind_nom_pens_ult1": "pensions",
+        "ind_recibo_ult1": "direct_debit",
+    }
+
+    # Rename columns
+    ds_train_renamed = ds_train.rename(col_names, axis=1)
+    ds_test_renamed = ds_test.rename(col_names, axis=1)
+
+    # Perform data type conversions
+    ds_train_renamed["age"] = pd.to_numeric(ds_train_renamed["age"], errors="coerce")
+    ds_train_renamed["gross_income"] = pd.to_numeric(
+        ds_train_renamed["gross_income"], errors="coerce"
+    )
+    ds_train_renamed["cust_seniority"] = pd.to_numeric(
+        ds_train_renamed["cust_seniority"], errors="coerce"
+    )
+    ds_train_renamed["cust_start_date_first_holder_contract"] = pd.to_datetime(
+        ds_train_renamed["cust_start_date_first_holder_contract"], errors="coerce"
+    )
+    ds_train_renamed["date_id"] = pd.to_datetime(ds_train_renamed["date_id"])
+
+    ds_test_renamed["age"] = pd.to_numeric(ds_test_renamed["age"], errors="coerce")
+    ds_test_renamed["gross_income"] = pd.to_numeric(
+        ds_test_renamed["gross_income"], errors="coerce"
+    )
+    ds_test_renamed["cust_seniority"] = pd.to_numeric(
+        ds_test_renamed["cust_seniority"], errors="coerce"
+    )
+    ds_test_renamed["cust_start_date_first_holder_contract"] = pd.to_datetime(
+        ds_test_renamed["cust_start_date_first_holder_contract"], errors="coerce"
+    )
+    ds_test_renamed["date_id"] = pd.to_datetime(ds_test_renamed["date_id"])
+
+    # Handle missing values
+    columns_to_drop = ["cust_last_primary_date", "spouse_index"]
+    ds_train_renamed.drop(
+        columns=[col for col in columns_to_drop if col in ds_train_renamed.columns],
+        axis=1,
+        inplace=True,
+    )
+    ds_test_renamed.drop(
+        columns=[col for col in columns_to_drop if col in ds_test_renamed.columns],
+        axis=1,
+        inplace=True,
+    )
+
+    cols = [
+        "emp_index",
+        "cust_country_res",
+        "cust_start_date_first_holder_contract",
+        "new_cust_index",
+        "cust_primary_type",
+        "cust_type_at_start_month",
+        "cust_rel_type_at_start_month",
+        "province",
+        "province_name",
+        "activity_index",
+        "channel_joined",
+        "cust_category",
+    ]
+
+    for col in cols:
+        if col in ds_train_renamed.columns and col in ds_test_renamed.columns:
+            ds_train_renamed[col].fillna(ds_train_renamed[col].mode()[0], inplace=True)
+            ds_test_renamed[col].fillna(ds_test_renamed[col].mode()[0], inplace=True)
```

Comment:
I believe the train/test split was done by time. In other words, we can assume test data is future data. So I think we should use the same values we got form the train set to fill nans in the test set because we should pretend not to know anything from the future (test) set. For example, imagine we are using live data. We wouldn't know what the mean/mode of values would be for the next month (test set). All we can do is to use the mean/mode form past months (test set) to impute missing values. We should do this for all imputing. in this file.

Reply:
Thanks, addressed within the notebook :)